sdk/src/loader_instruction.rs

```diff
@@ -1,6 +1,8 @@
 use crate::instruction::{AccountMeta, Instruction};
 use crate::pubkey::Pubkey;
 use crate::sysvar::rent;
+use bincode::serialize;
+use serde::Serialize;
 
 #[derive(Serialize, Deserialize, Debug, PartialEq, Eq, Clone)]
 pub enum LoaderInstruction {
```

Comment:
We should switch to short_vec for the `bytes` vectors in each of these instructions.  The full 64bit vec is wasteful, and opens up the implementation to $5 🔧 attacks by passing in a vector length of 2^64

Reply:
That is also a bummer because `ShortVec` is very inefficient at serializing bytes.  Why don't we use bincode limits to protect ourselves instead of `ShortVec`?

Reply:
What makes `ShortVec` so inefficient?  We use it in every transaction so seems if it's so bad then that'd be a super thing to optimize regardless

Reply:
These are very bad for byte arrays:
https://github.com/solana-labs/solana/blob/929a81e636fc4568257c5d0668e0008d77ef437b/sdk/src/short_vec.rs#L105
https://github.com/solana-labs/solana/blob/929a81e636fc4568257c5d0668e0008d77ef437b/sdk/src/short_vec.rs#L135

Why don't we dump `ShortVect` in leu of https://docs.rs/bincode/1.0.0-alpha7/bincode/trait.SizeLimit.html.  Prob should set the limit down around packet size.

Reply:
We use ShortVec primarily to avoid consuming 8 bytes for a vector length when 1 byte will do in most cases.  This reduces the over-the-wire size of a Transaction.

Reply:
Yeah I know, we save 5 bytes by cutting the length down to 3 bytes.  It looks like equally, we are using it to protect against DDoS by limiting the size of the vector.  We should probably be doing that explicitly with bincode limits and optimizing ShortVec for byte arrays.

Reply:
@mvines   I'm planning to commit this PR without using `ShortVec` for `LoaderInstruction` vectors.  Instead, I'm tracking with this issue: https://github.com/solana-labs/solana/issues/6517

Also, I'm a little confused about how not using `ShortVec` in the `LoaderInstruction` could lead to a DDoS.  Loader instructions are not read from an end-less stream; they are deserialized from packets of known size and therefore are already limited to a smaller size than the `u16` that `ShortVec` imposes.  Am I missing another attack angle?

Reply:
If the `Vec` deserialization looks like `ShortVec` then it ends up allocating the vector to the input size up front, based on the length field directly regardless of how many octets of data follow.  eg:

https://github.com/solana-labs/solana/blob/01fe7c90a554333e99d834ef1fc5f2c3832dacd5/sdk/src/short_vec.rs#L134

Reply:
Thanks, sounds like bincode byte limits are the correct answer for DDoS prevention and a better solution than `ShortVec` since we can limit the deserialization size down to packet size (or we add byte limits to `ShortVec`): https://docs.rs/bincode/1.2.0/bincode/struct.Config.html#byte-limit-details

Reply:
Cool.  Except that bincode wastes 7 bytes for the full Vec length when in most cases we only need 1 byte

Reply:
Yup, plan to use binode limits and look further into optimizaing ShortVec